kernel/page_allocator/src/lib.rs

```diff
@@ -429,38 +451,56 @@ fn find_specific_chunk(
 
 /// Searches the given `list` for any chunk large enough to hold at least `num_pages`.
 ///
-/// It first attempts to find a suitable chunk **not** in the designated regions,
+/// If a given range is specified, the returned `AllocatedPages` *must* exist
+/// fully within that inclusive range of pages.
+///
+/// If no range is specified, this function first attempts to find a suitable chunk
+/// that is **not** within the designated regions,
 /// and only allocates from the designated regions as a backup option.
 fn find_any_chunk(
 	list: &mut StaticArrayRBTree<Chunk>,
-	num_pages: usize
+	num_pages: usize,
+	within_range: Option<&PageRange>,
 ) -> Result<(AllocatedPages, DeferredAllocAction<'static>), AllocationError> {
-	let designated_low_end = DESIGNATED_PAGES_LOW_END.get().ok_or(AllocationError::NotInitialized)?;
+	let designated_low_end = DESIGNATED_PAGES_LOW_END.get()
+		.ok_or(AllocationError::NotInitialized)?;
+	let full_range = PageRange::new(*designated_low_end + 1, DESIGNATED_PAGES_HIGH_START - 1);
+	let range = within_range.unwrap_or(&full_range);
 
-	// During the first pass, we ignore designated regions.
+	trace!("Allocating {} pages within range {:?}", num_pages, range);
+
+	// During the first pass, we only search within the given range.
+	// If no range was given, we search from the end of the low designated region
+	// to the start of the high designated region.
 	match list.0 {
 		Inner::Array(ref mut arr) => {
 			for elem in arr.iter_mut() {
 				if let Some(chunk) = elem {
-					// Skip chunks that are too-small or in the designated regions.
-					if  chunk.size_in_pages() < num_pages || 
-						chunk.start() <= designated_low_end || 
-						chunk.end() >= &DESIGNATED_PAGES_HIGH_START
-					{
-						continue;
-					} 
-					else {
-						return adjust_chosen_chunk(*chunk.start(), num_pages, &chunk.clone(), ValueRefMut::Array(elem));
+					let lowest_start_page = *max(chunk.start(), range.start());
+					let highest_end_page  = *min(chunk.end(), range.end());
```

Comment:
I think it should be
```suggestion
					let lowest_start_page = *min(chunk.start(), range.start());
					let highest_end_page  = *max(chunk.end(), range.end());
```

Reply:
Hmm. My logic here is that the the range of pages we end up allocating needs to be within _both_ the chunk and the range. Looking at the first one, the lowest possible start page must be within both the chunk and the specified range, meaning that it has to be greater than both the chunk's start page and range's start part. If we use `min` for `lowest_start_page`, the Chunk's starting page might be below (outside of) the given range, or the range's starting page might also be below the chunk. So that wouldn't work.

The same is true for the highest-possible ending page, just inverted. That seems correct to me; did I make a mistake?

Reply:
Yea ok you're right. I think the variable names also tripped me up because the lowest start page is actually the largest of the two start pages :), similarly for the end page.

Reply:
Ah I see. Will add some comments to that effect